blobstore/src/main/java/org/jclouds/blobstore/util/BlobStoreUtils.java

```diff
@@ -85,4 +103,65 @@ private static void convertUserMetadataKeysToLowercase(MutableBlobMetadata metad
       }
       metadata.setUserMetadata(lowerCaseUserMetadata);
    }
+
+   private static class CrawlBlobStoreIterable implements Iterable<StorageMetadata> {
+      private final BlobStore blobStore;
+      private final String containerName;
+      private final ListContainerOptions options;
+
+      CrawlBlobStoreIterable(BlobStore blobStore, String containerName, ListContainerOptions options) {
+         this.blobStore = checkNotNull(blobStore);
+         this.containerName = checkNotNull(containerName);
+         this.options = checkNotNull(options);
+      }
+
+      @Override
+      public Iterator<StorageMetadata> iterator() {
+         return new CrawlBlobStoreIterator(blobStore, containerName, options);
+      }
+   }
+
+   private static class CrawlBlobStoreIterator implements Iterator<StorageMetadata> {
+      private final BlobStore blobStore;
+      private final String containerName;
+      private final ListContainerOptions options;
+      private Iterator<? extends StorageMetadata> iterator;
+      private String marker;
+
+      CrawlBlobStoreIterator(BlobStore blobStore, String containerName, ListContainerOptions options) {
+         this.blobStore = checkNotNull(blobStore);
+         this.containerName = checkNotNull(containerName);
+         this.options = checkNotNull(options).recursive();
+         advance();
+      }
+
+      private void advance() {
+         if (marker != null) {
+            options.afterMarker(marker);
+         }
+         PageSet<? extends StorageMetadata> set = blobStore.list(containerName, options);
+         marker = set.getNextMarker();
+         iterator = set.iterator();
+      }
+
+      @Override
+      public boolean hasNext() {
+         return iterator.hasNext() || marker != null;
+      }
+
+      @Override
+      public StorageMetadata next() {
+         while (true) {
+            if (!iterator.hasNext()) {
+               advance();
+            }
+            StorageMetadata metadata = iterator.next();
```

Comment:
To clarify: it is unlikely that the provider returns an empty page if there is a marker, but since it is quite a common pattern to use fallbacks that return empty lists on 404, we should take care and be defensive here. Don't assume that the iterable has content at this point.

Also IMO this class deserves its own unit test to verify it behaves properly when there are 0 pages (the initial advance returns nothing), a single page, multiple pages, and empty pages.